### ISSUE TITLE
Constructor not called when `from_attributes=True` is used

### ISSUE THREAD

**Issue by aguckenber-chwy** [pydantic/pydantic#8142]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Had some V1 code that I was porting to V2 and noticed a bug with PrivateAttrs set in the constructor of a model.  The constructor is not consistently called. Specifically if a random model is passed to `model_validate()` which can be parsed with `from_attributes`. See example for more details.

### Example Code

```Python
from pydantic import BaseModel, ConfigDict, PrivateAttr

class Foo:
    def __init__(self):
        self.foo = "baz"

class Example(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    foo: str = "bar"
    _priv: str = PrivateAttr()

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        print("here")
        self._priv = "password"

# e = Example() # This enters the constructor
# e = Example.model_validate({"foo": "bip"}) # This enters the constructor
e = Example.model_validate(Foo()) # This does not enter the constructor
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.5.0
        pydantic-core version: 2.14.1
          pydantic-core build: profile=release pgo=false
                 install path: /workspace/app/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.13 (main, Sep 20 2023, 21:37:20) [GCC 10.2.1 20210110]
                     platform: Linux-6.3.13-linuxkit-aarch64-with-glibc2.31
             related packages: fastapi-0.103.2 mypy-0.910 pydantic-settings-2.1.0 typing_extensions-4.8.0
```

**Comment by sydney-runkle** [comment]
Hi @aguckenber-chwy,

Thanks for reporting this issue. This does seem like a bug, likely on the `pydantic-core` side of things 🐛. We'll look into a fix to include in our next minor release 👍.

**Comment by aguckenber-chwy** [comment]
> Thought this was a feature, not a bug. Something like this:
> 
> ```python
> class Ok(BaseModel, t.Generic[T]):
>     result: T
>     ok: t.Literal[True] = Field(True, kw_only=True)
> 
>     def __init__(self, result: T) -> None:
>         super().__init__(result=result)
> ```

I'm confused what your example is supposed to show? It doesn't appear to relate to the original post/bug at all?

**Comment by rafales** [comment]
Yeah, I guess that was a little bit confusing.

I'm customizing `__init__` in a way that would break if functions like `model_validate()` would call it (taking one argument, as a positional one). It made sense to me, as this is pretty similar to what SQLAlchemy is doing with models (you can use `__init__` to construct objects, but when data is read from database, `__init__` is not called).

@sydney-runkle said that it seemed like a bug, so I pointed out that it may be used as a feature as well by some folks.

@aguckenber-chwy as for your case, maybe overriding `model_post_init` is what you need?

**Comment by aguckenber-chwy** [comment]
@rafales  The reason its filed under a bug is because it breaks specifically how `PrivateAttr`s are supposed to work, without the constructor, you can't create a private attribute based on a users input.

SQLAlchemy does this a little different by using `__new__` when reading from a database  but also works a little differnet. When SQLAlchemy reads from a database it expects the data to match the model and thus does not call the models initializer. I.e. it call `__new__` but never calls `__init__`.For SQLAlchemy this make a little sense though its worth noting they add `@reconstructor` to work around the fact that they do not call the initializer.

Pydantic is different, specifically with `model_validate`. I am asking Pydantic to create a model based on the object I pass it, I am not saying the object I pass it directly maps to the model, I just want pydantic to rip attributes from it. I.e. I am using `from_attributes`, not `from_identical_model`l.

In theory, if the stance pydantic wants to take is the SQLAlchemy approach, they can change the docs to have people overwrite `__new__` instead of `__init__` for `PrivateAttr`s though I would argue this is still a bug and falls under a violation of the least surprise principal. If I call a function with three different arguments, I wouldn't expect 2 instances to call the initializer and one not to. If its a surprise, its a bug. No one likes to read the docs and that say "if you do ... than X BUT if you do ... then its Y or possible even Z if you do ...".

Pydantic is not an ORM, it should be calling the initializer (`__init__`).

**Comment by Bernardoow** [comment]
Hey I have the same probleman. I fixed when I pass the variable from_attributes to model_validation .

`.model_validate(obj, from_attributes=True)`

**Comment by nix010** [comment]
@sydney-runkle I can take this one. My mind-map of `pydantic-core` is still fresh.

**Comment by sydney-runkle** [comment]
@nix010,

Great! Assigned to you :).

**Comment by nix010** [comment]
@sydney-runkle @aguckenber-chwy Got issue while working on this. What do you think is expect result for this case ?
We want the `Example.__init__()` get called but with what args and kwargs ?

```python

from pydantic import BaseModel, ConfigDict


class Foo:
    foo = 'bar1'
    def __init__(self):
        self.foo = "baz_new"

class Example(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    foo: str = "bar"
    key_a: str = "a"
    key_b: str = "b"

    def __init__(self, a, b):
        self.key_a = a
        self.key_b = b

# This raised error
# e = Example.model_validate({"foo": "asdas"})

# But what args, kwargs should this __init__ be called with ?
e = Example.model_validate(Foo())

```

**Comment by sydney-runkle** [comment]
Obviously, given the bumps above, this is low priority (only 2 likes on the main description doesn't suggest much demand).

Bumping this off of the milestone - there's more to figure out here. Specifically:

You can't call `__init__` with `from_attributes` like you can with `model_validate`. `__init__` inherently accepts a list of args/kwargs corresponding to model attributes. As @nix010 raised above, the input to the custom `__init__` is unclear - should we use the `__dict__` of the input instance? That's error prone and we run into issues with extra.

**Comment by gerhardadler** [comment]
@sydney-runkle

Maybe making `__init__` and `model_validate` completely seperate is a solution to consider?

The user would still have the option to override `model_validate` seperately if custom behaviour is needed.

I think this would make the behavior more consistent.

It would also solve this issue: https://github.com/pydantic/pydantic/issues/10062

**Comment by sydney-runkle** [comment]
Yeah, I understand the value here conceptually, but that's a really big change internally. Probably not doable in the short term.
